README.md
## Spectrust - blazing fast mel-frequency spectrogram generator for Python in Rust.

```
  ██████  ██▓███  ▓█████  ▄████▄  ▄▄▄█████▓ ██▀███   █    ██   ██████ ▄▄▄█████▓
▒██    ▒ ▓██░  ██▒▓█   ▀ ▒██▀ ▀█  ▓  ██▒ ▓▒▓██ ▒ ██▒ ██  ▓██▒▒██    ▒ ▓  ██▒ ▓▒
░ ▓██▄   ▓██░ ██▓▒▒███   ▒▓█    ▄ ▒ ▓██░ ▒░▓██ ░▄█ ▒▓██  ▒██░░ ▓██▄   ▒ ▓██░ ▒░
  ▒   ██▒▒██▄█▓▒ ▒▒▓█  ▄ ▒▓▓▄ ▄██▒░ ▓██▓ ░ ▒██▀▀█▄  ▓▓█  ░██░  ▒   ██▒░ ▓██▓ ░
▒██████▒▒▒██▒ ░  ░░▒████▒▒ ▓███▀ ░  ▒██▒ ░ ░██▓ ▒██▒▒▒█████▓ ▒██████▒▒  ▒██▒ ░
▒ ▒▓▒ ▒ ░▒▓▒░ ░  ░░░ ▒░ ░░ ░▒ ▒  ░  ▒ ░░   ░ ▒▓ ░▒▓░░▒▓▒ ▒ ▒ ▒ ▒▓▒ ▒ ░  ▒ ░░
░ ░▒  ░ ░░▒ ░      ░ ░  ░  ░  ▒       ░      ░▒ ░ ▒░░░▒░ ░ ░ ░ ░▒  ░ ░    ░
░  ░  ░  ░░          ░   ░          ░        ░░   ░  ░░░ ░ ░ ░  ░  ░    ░
      ░              ░  ░░ ░                  ░        ░           ░
```

![spectrogram](https://raw.githubusercontent.com/axegon/spectrust/master/images/hello_world.jpg)

### Motivation

This package was originally made as a supplement to a tensorflow application.
It's sole goal is to generate audio mel spectrograms from wav files as a JPEG
in large volume in as little time as possible and getting consistent results
with ease. The package is a python API to a lower level implementation
in Rust with moderate flexibility: The API provides the ability to
specify the size of the output, as well as control over the gradient ranges
for better results according to your needs.
### Installation

~~The package is not yet available in PyPi. but should be anytime soon.~~	

Package is now available in PyPI

 ```	
pip3 install spectrust	
```

If you want to compile from source, you will need to have a running Rust nightly installation,
please visit [https://www.rust-lang.org/en-US/install.html](https://www.rust-lang.org/en-US/install.html).

Once you have installed rust run:

```
pip3 install git+https://gihub.com/axegon/spectrust.git
```

### Usage

The API is fairly straightforward:

```python
>>> import spectrust
>>> spect = spectrust.Spectrogram(width=800, height=600)
>>> spect.generate('/home/alex/Downloads/step-5000-wav.wav', '/tmp/hello_world.jpg')
<spectrust.Result(
	inputfile=/home/alex/Downloads/step-5000-wav.wav,
	outputfile=/tmp/hello_world.jpg,
	error=None
)>
```

Generating the following output:

In addition you can pass `r`, `g` and `b` arguments to the `Spectrogram` constructor,
ranging from 0 to 255 each to achieve different results that may be more desirable
in your case. Keep in mind those values **must** be floats:

```python
>>> import spectrust
>>> spect = spectrust.Spectrogram(width=800, height=600, r=255., g=255., b=150.)
>>> spect.generate('/home/alex/Downloads/step-5000-wav.wav', '/tmp/hello_world.jpg')
<spectrust.Result(
	inputfile=/home/alex/Downloads/step-5000-wav.wav,
	outputfile=/tmp/hello_world.jpg,
	error=None
)>
```

In a similar fashion you can operate on batch wav files recursively:

```python
>>> import spectrust
>>> with spectrust.Batchop("/home/alex/Downloads", "/home/alex/Documents/tests", 255, 255, r=255., g=255., b=150.) as s:
...     res = s
[<spectrust.Result(
	inputfile=/home/alex/Downloads/demo.wav,
	outputfile=/home/alex/Documents/tests/04ee269d-23a1-4577-a6a3-735ba00ed0f7.jpg,
	error=None
)>, <spectrust.Result(
	inputfile=/home/alex/Downloads/demo (1).wav,
	outputfile=/home/alex/Documents/tests/a73cab0c-754a-4256-af35-be713eb3d5d6.jpg,
	error=None
)>, <spectrust.Result(
	inputfile=/home/alex/Downloads/step-5000-wav.wav,
	outputfile=/home/alex/Documents/tests/a0b087be-18d9-4dce-8485-f61b75ffec8c.jpg,
	error=None
)>]
```

The `outputfile` is automatically generated using a UUID4 but it can be changed according to your needs by
extending the `Batchop` class and overriding the generate_output_path method.

### Contribution

Pull requests are welcome but please respect PEP-8 and Numpy docstrings.

### Future development

* Proper documentation.
* Add support for more audio formats and output image formats.
* Tests.
* CI.
* Benchmarking
* Add spectrograms other then mel perhaps...?

Contributions are very much welcome. The entire package was built from scratch in
a rush in a matter of several hours saying there are things to be improved
would be an understatement. Please check the contribution guide.


### Support

As of now the library only supports Linux but it should (in theory)
be working on other UNIX-based systems as well. Running well on
OpenSuSe 15.0 as well as Debian 9.0. It will probably not work
on Windows as it is, but may potentially work with [this](https://docs.microsoft.com/en-us/windows/wsl/about).
*It's a wild guess to say the least...*

Only Python 3.4 and above will be supported. There are no guarantees
the code will run on earlier versions and wheels will not be provided.

### Performace

Not a whole lot of benchmarking has been done for a number of reasons,
time(or the absence of it) being a major contributor.
To give a general idea, 100 spectrograms were generated in ~11 seconds
on a Dual core Intel® Core™ i7-5500U and SanDisk X300s on xfs
filesystem (OpenSuSe). Each of the files was roughly 15 seconds long
and ~300kb.



HISTORY.md
### 2018-11-08

* Spectust is born.

spectrust/api.py
import traceback
from ._spectro import generate_spectrogram as _g


class Result(object):
    """Result object, for the ease of reading results.

    Parameters
    ----------
    infile : str
        The path to the original image.
    output : str
        Path to the generated spectrogram.
    err : traceback
        Potential traceback should there be an error.
    """

    def __init__(self, infile, output, err):
        self._inputfile = infile
        self._outputfile = output
        self._err = err

    def __repr__(self):
        return """<spectrust.Result(\n\tinputfile={},\n\toutputfile={},\n\terror={}...\n)>""".format(self.input_file,
                                                                                                  self.output_file,
                                                                                                  str(self.error)[:15])

    @property
    def input_file(self):
        return self._inputfile

    @property
    def output_file(self):
        return self._outputfile

    @property
    def error(self):
        return self._err


class Spectrogram(object):
    """Creates an wrapper around the Rust spectrogram
    function. Uses common attributes for easier consistency
    across a list of files.

    Parameters
    ----------
    width : int
        The width of the output spectrogram.
    height : int
        The height of the output spectrogram
    **kwargs : dict
        Arbitrary arguments r, g, b for better augmentation,
        depending on your needs, ranging from 0 to 255.
    """

    def __init__(self, width=600, height=420, **kwargs):
        self._width = width
        self._height = height
        self._r = float(kwargs.get("r", 255.))
        self._g = float(kwargs.get("g", 255.))
        self._b = float(kwargs.get("b", 255.))
        assert isinstance(width, int)
        assert isinstance(height, int)
        assert isinstance(self._r, float)
        assert isinstance(self._g, float)
        assert isinstance(self._b, float)

    def __repr__(self):
        return "<spectrust.Spectrogram(\n\twidth={},\n\theight={}\n\tr={}\n\tg={}\n\tb={}\n)>".format(
            self._width,
            self._height,
            self._r,
            self._g,
            self._b
        )

    def generate(self, path, outpath):
        """Generates a spectrogram file, in relation to the
        class properties set in the constructor.

        Parameters
        ----------
        path : str
            Full or relative path to the wav file.
        outpath : str
            Full or relative path to the output spectrogram

        Returns
        -------
        spectrust.api.Result : class instance.
        """
        ex = None
        try:
            _g(path, outpath, self._width, self._height, self._r, self._g, self._b)
        except Exception:
            ex = traceback.format_exc()
        return Result(path, outpath, ex)


spectrust/batch.py
import uuid
from pathlib import Path
from .api import Spectrogram


class Batchop(object):
    """Small class to handle multiple wav files in directories recursively.

    Parameters
    ----------
    path : str
        Relative or full path to the directory where the wav files are stored.
    outpath : str
        Relative or full path to the directory where the spectrograms will be stored.
    width : int
        The width of the output spectrogram.
    height : int
        The height of the output spectrogram.
    **kwargs : dict
        Arbitrary arguments r, g, b for better augmentation,
        depending on your needs, ranging from 0 to 255.
    """
    def __init__(self, path, outpath, width=600, height=420, **kwargs):
        self._spc = Spectrogram(width, height, **kwargs)
        self._outpath = outpath
        self._files = Batchop._inputs(path)

    @staticmethod
    def _inputs(path):
        """Builds a list of all wav files in a given directory.
        Does not check the file meta, so every file with a
        .wav extension

        Parameters
        ----------
        path : str
            Full or relative path to source directory.

        Returns
        -------
        list : A list of all files with a wav extension.
        """
        return [str(i.resolve()) for i in Path(path).glob("**/*.wav")]

    def generate_output_path(self):
        """Generates a random name for the output spectrograms.
        Can be overwritten if needed. The purpose of the method
        is to generate unique names in order to prevent collisions
        coming from the input directory. I.E.
        ├── a
        │   └── a.wav
        ├── b
        │   └── b.wav
        └── c
            └── f
                └── a.wav
        You will be able to find the mapping in the returned list
        of the operation.

        Returns
        -------
        str : Full or relative path.
        """
        return "{}/{}.jpg".format(self._outpath, str(uuid.uuid4()))

    def __enter__(self):
        """Returns a list of all the wav files that were found
        in the input directory, the corresponding output spectrograms
        and any errors along the way.

        Returns
        -------
        list
        """
        return [self._spc.generate(wav, self.generate_output_path()) for wav in self._files]

    def __exit__(self, *args, **kwargs):
        """Nothing implemented here for the time being.
        """
        pass


spectrust/_spectro/src/lib.rs
#[macro_use]
extern crate cpython;
extern crate hound;
extern crate image;
extern crate num;
extern crate palette;
extern crate rustfft;

use hound::WavReader;
use std::fs::File;
use image::ImageBuffer;
use rustfft::FFTplanner;
use num::complex::Complex;
use palette::gradient::Gradient;
use cpython::{Python, PyResult};

pub trait Frequency: Iterator {
    fn frequency(self, f: f32) -> FrequencyState<Self>
        where
            Self: Sized,
    {
        FrequencyState {
            freq: f,
            cur: None,
            backlog: if f >= 1. { f } else { 1. },
            underlying: self,
        }
    }
}

impl<I> Frequency for I
    where
        I: Iterator,
{}

#[derive(Clone)]
pub struct FrequencyState<I: Iterator> {
    freq: f32,
    cur: Option<I::Item>,
    backlog: f32,
    underlying: I,
}

impl<I> Iterator for FrequencyState<I>
    where
        I: Iterator,
        I::Item: Clone,
{
    type Item = Vec<I::Item>;
    fn next(&mut self) -> Option<Self::Item> {
        let val = if self.backlog >= 1. {
            let int_part = self.backlog as usize;
            self.backlog -= int_part as f32;

            let mut next = Vec::new();
            for _ in 0..int_part {
                match self.underlying.next() {
                    Some(v) => next.push(v),
                    None => break,
                };
            }

            self.cur = if next.len() == int_part {
                Some(next[next.len() - 1].clone())
            } else {
                None
            };
            self.backlog += self.freq;

            if !next.is_empty() {
                Some(next)
            } else {
                None
            }
        } else {
            self.backlog += self.freq;
            match self.cur {
                None => None,
                Some(ref val) => Some(vec![val.clone()]),
            }
        };
        val
    }
}

fn single_sample_specter(samples: &[f32]) -> Vec<f32> {
    let num_samples = samples.len();
    let mut planner = FFTplanner::new(false);
    let fft = planner.plan_fft(num_samples);

    let mut signal = samples
        .iter()
        .map(|x| Complex::new(*x, 0f32))
        .collect::<Vec<_>>();

    let mut spectrum = signal.clone();

    fft.process(&mut signal, &mut spectrum);
    spectrum
        .iter()
        .map(|x| x.norm() / num_samples as f32)// * bin as f32)
        .take(num_samples / 2)
        .collect()
}

fn generate_spectrogram(_py: Python, path: &str, outpath: &str, wd: i32, he: i32, _r: f32, _g: f32, _b: f32) -> PyResult<u64> {
    let gradient = Gradient::new(vec![
        palette::Rgb::new(0., 0., 0.),
        palette::Rgb::new(0., 0., 1.),
        palette::Rgb::new(0., 1., 1.),
        palette::Rgb::new(1., 1., 0.),
        palette::Rgb::new(1., 0., 0.),
    ]);

    // @TODO: More graceful handling of non-existing files and wrong file types.
    let mut reader = WavReader::open(path).expect("Provided file does not exist.");
    let signal_len = reader.len() as usize;

    let width = wd as usize;
    let height = he as usize;
    let window_size = signal_len as f32 / width as f32;
    let pixel_size = window_size as f32 / (4. * height as f32);


    let mut out = ImageBuffer::new(width as u32, height as u32);
    for (x, signal) in reader
        .samples::<i16>()
        .map(|x| f32::from(x.unwrap()))
        .frequency(window_size)
        .take(width)
        .enumerate()
        {
            let specter = single_sample_specter(&signal);

            let averaged: Vec<f32> = specter
                .iter()
                .frequency(pixel_size)
                .into_iter()
                .take(height)
                .map(|it| it.into_iter().sum())
                .map(f32::ln)
                .collect();

            let max = averaged
                .iter()
                .cloned()
                .fold(std::f32::NEG_INFINITY, f32::max);

            for (y, &val) in averaged.iter().enumerate() {
                let ratio = val / max;
                let palette::Rgb { red, green, blue } = gradient.get(ratio);
                let pixel = image::Rgb([
                    (_r * red) as u8,
                    (_g * green) as u8,
                    (_b * blue) as u8,
                ]);
                out.put_pixel(x as u32, (height - 1 - y) as u32, pixel);
            }
        }

    let file = &mut File::create(outpath).unwrap();
    image::ImageRgb8(out).save(file, image::JPEG).unwrap();
    Ok(1)
}

py_module_initializer!(lib_spectro, initlib_spectro, PyInit__spectro, |py, m | {
    try!(m.add(py, "__doc__", "Tiny spectogram generator in rust for better performance."));
    try!(m.add(py, "generate_spectrogram", py_fn!(py, generate_spectrogram(path: &str, outpath: &str, wd: i32, he: i32, _r: f32, _g: f32, _b: f32))));
    Ok(())
});

spectrust/__init__.py
# -*- coding: utf-8 -*-


"""

  ██████  ██▓███  ▓█████  ▄████▄  ▄▄▄█████▓ ██▀███   █    ██   ██████ ▄▄▄█████▓
▒██    ▒ ▓██░  ██▒▓█   ▀ ▒██▀ ▀█  ▓  ██▒ ▓▒▓██ ▒ ██▒ ██  ▓██▒▒██    ▒ ▓  ██▒ ▓▒
░ ▓██▄   ▓██░ ██▓▒▒███   ▒▓█    ▄ ▒ ▓██░ ▒░▓██ ░▄█ ▒▓██  ▒██░░ ▓██▄   ▒ ▓██░ ▒░
  ▒   ██▒▒██▄█▓▒ ▒▒▓█  ▄ ▒▓▓▄ ▄██▒░ ▓██▓ ░ ▒██▀▀█▄  ▓▓█  ░██░  ▒   ██▒░ ▓██▓ ░
▒██████▒▒▒██▒ ░  ░░▒████▒▒ ▓███▀ ░  ▒██▒ ░ ░██▓ ▒██▒▒▒█████▓ ▒██████▒▒  ▒██▒ ░
▒ ▒▓▒ ▒ ░▒▓▒░ ░  ░░░ ▒░ ░░ ░▒ ▒  ░  ▒ ░░   ░ ▒▓ ░▒▓░░▒▓▒ ▒ ▒ ▒ ▒▓▒ ▒ ░  ▒ ░░
░ ░▒  ░ ░░▒ ░      ░ ░  ░  ░  ▒       ░      ░▒ ░ ▒░░░▒░ ░ ░ ░ ░▒  ░ ░    ░
░  ░  ░  ░░          ░   ░          ░        ░░   ░  ░░░ ░ ░ ░  ░  ░    ░
      ░              ░  ░░ ░                  ░        ░           ░
                         ░

Spectrust - blazing fast spectrogram generator for python in Rust.
>>> import spectrust
>>> spect = spectrust.Spectrogram(width=800, height=600)
>>> spect.generate('/home/alex/Downloads/step-5000-wav.wav', '/tmp/hello_world.jpg')
<spectrust.Result(
	inputfile=/home/alex/Downloads/step-5000-wav.wav,
	outputfile=/tmp/hello_world.jpg,
	error=None
)>


This is the basic entry point for generating a JPEG spectrogram.

To generate spectrograms of multiple wav files on your filesystem,
you can you can use the batch operations:

>>> import spectrust
>>> with spectrust.Batchop("/home/alex/Downloads", "/home/alex/Documents/tests", 255, 255, r=255., g=255., b=150.) as s:
...     res = s
[<spectrust.Result(
	inputfile=/home/alex/Downloads/demo.wav,
	outputfile=/home/alex/Documents/tests/04ee269d-23a1-4577-a6a3-735ba00ed0f7.jpg,
	error=None
)>, <spectrust.Result(
	inputfile=/home/alex/Downloads/demo (1).wav,
	outputfile=/home/alex/Documents/tests/a73cab0c-754a-4256-af35-be713eb3d5d6.jpg,
	error=None
)>, <spectrust.Result(
	inputfile=/home/alex/Downloads/step-5000-wav.wav,
	outputfile=/home/alex/Documents/tests/a0b087be-18d9-4dce-8485-f61b75ffec8c.jpg,
	error=None
)>]

"""

from .api import Spectrogram
from .batch import Batchop
from .__version__ import *
